### PR TITLE
replyPop lemmas in Finalise_R

### PR DESCRIPTION
**Pull request by VictorPhan**
Rebased and cleaned up work for replyPop_invs'.
Ready to be reviewed, but yet to add fixes to broken proofs in later theory files due to these changes.

### CODE REVIEW

File: spec/haskell/src/SEL4/Object/Reply.lhs
Comment (VictorPhan) [inline spec/haskell/src/SEL4/Object/Reply.lhs:83]: spec update for replyPop, to be removed when @mktnk3's [PR](https://github.com/seL4/l4v/pull/185) is merged

File: proof/refine/ARM/Finalise_R.thy
Comment (VictorPhan) [inline proof/refine/ARM/Finalise_R.thy:2425]: These lemmas use `schedContextDonate` lemmas so we need to wait for @corlewis's [PR](https://github.com/seL4/l4v/pull/245) moving these lemmas up before merging.

File: proof/refine/ARM/Finalise_R.thy
Comment (VictorPhan) [inline proof/refine/ARM/Finalise_R.thy:2561]: @MitchellBuckley this is the lemma we discussed you should take a look at

**Comment by mktnk3** [comment]
It looks like there are several files that shouldn’t be included in the PR. Some kind of temp/backup files. You need to remove them.

File: proof/refine/ARM/Ipc_R.thy
Comment (MitchellBuckley) [inline proof/refine/ARM/Ipc_R.thy:2832]: It seems unusual that so many of these were added in, is there an innocent reason for this?

File: proof/refine/ARM/Finalise_R.thy
Comment (MitchellBuckley) [inline proof/refine/ARM/Finalise_R.thy:4299]: Is there a plan with this sorry? (And some similar sorries which don't have any comments attached to them).

File: proof/refine/ARM/Finalise_R.thy
Comment (MitchellBuckley) [inline proof/refine/ARM/Finalise_R.thy:2810]: Excellent! Nitpick: please move this up to Invariants_H or similar.

File: proof/refine/ARM/Finalise_R.thy
Comment (mktnk3) [inline proof/refine/ARM/Finalise_R.thy:2417]: Nice!

File: proof/refine/ARM/Finalise_R.thy
Comment (mktnk3) [inline proof/refine/ARM/Finalise_R.thy:2548]: Is this lemma name what it is supposed to be?

File: proof/refine/ARM/Finalise_R.thy
Comment (mktnk3) [inline proof/refine/ARM/Finalise_R.thy:2327]: Will this be the last sorry in Finalise_R?

**Review by mktnk3 — COMMENTED** [review]
I haven't quite finished reviewing yet, but I'll send the comments so far for now. Will continue tomorrow.

File: proof/refine/ARM/Finalise_R.thy
Comment (VictorPhan) [reply proof/refine/ARM/Finalise_R.thy:2548]: Oh I haven't updated the name for this one after creating it. Maybe `updateReply_obj_at'_sc_ksIdleThread`?

File: proof/refine/ARM/Ipc_R.thy
Comment (VictorPhan) [reply proof/refine/ARM/Ipc_R.thy]: Was just temporary, it's removed now

File: proof/refine/ARM/Finalise_R.thy
Comment (VictorPhan) [reply proof/refine/ARM/Finalise_R.thy]: This is beyond the scope of my PR, but it changed because of the replyPop spec tweaks.

File: proof/refine/ARM/Finalise_R.thy
Comment (VictorPhan) [reply proof/refine/ARM/Finalise_R.thy:2336]: This lemma is unused, so I'll remove it instead

File: proof/refine/ARM/Finalise_R.thy
Comment (VictorPhan) [reply proof/refine/ARM/Finalise_R.thy:2341]: Unused; removed.

File: proof/refine/ARM/Finalise_R.thy
Comment (VictorPhan) [reply proof/refine/ARM/Finalise_R.thy:2359]: Removed; replaced.

File: proof/refine/ARM/Finalise_R.thy
Comment (VictorPhan) [reply proof/refine/ARM/Finalise_R.thy:2327]: There is also `replyPop_iflive` and `replyRemove_corres` left

File: proof/refine/ARM/Finalise_R.thy
Comment (VictorPhan) [reply proof/refine/ARM/Finalise_R.thy:2379]: It's unused, but since it's generic enough, I'll leave it in (and move it to Lib).

File: proof/refine/ARM/Finalise_R.thy
Comment (mktnk3) [reply proof/refine/ARM/Finalise_R.thy:2548]: yes, that sounds good!

File: proof/refine/ARM/Finalise_R.thy
Comment (MitchellBuckley) [reply proof/refine/ARM/Finalise_R.thy:2546]: I believe that Michael took a look at this. I suppose his work should be cherry-picked on top here?

**Review by MitchellBuckley — APPROVED** [review]
Looks nice!
I assume that the single added sorry will be resolved before merging, or perhaps it will stay in here and be removed shortly afterwards with a separate PR?

File: proof/refine/ARM/Finalise_R.thy
Comment (VictorPhan) [reply proof/refine/ARM/Finalise_R.thy]: Was a quick fix so I'm it in this PR

File: proof/refine/ARM/Finalise_R.thy
Comment (VictorPhan) [inline proof/refine/ARM/Finalise_R.thy:3760]: will change this line back

**Comment by VictorPhan** [comment]
Commits are cleaned up now, @michaelmcinerney will handle merging with his cherry-picked commit after he rebases his awaken PR.

File: proof/refine/ARM/Finalise_R.thy
Comment (mktnk3) [reply proof/refine/ARM/Finalise_R.thy:2327]: Michael took care of the first and the latter is a temporary sorry for what Corey is currently working on in IpcCancel_R.

**Pull request closed without merging**
